### PR TITLE
Instructor: students: add invite button #8648

### LINKED ISSUE #8648: Instructor: students: add invite button
**Environment**: 
`master` branch commit at 497bf6d

Here's the screenshot for InstructorCourseDetailsPage:
image

Here's the screenshot for InstructorStudentListPage
image

Instructors are likely to go to the Students page to look at the students, and send invite email if they have not joined the course yet. However, the button to send invite is missing.

### PR DESCRIPTION
**Pull request by nashmia-riaz**
Fixes #8648

Simply added "or" condition in line 137 of file studentList.tag to allow the button to appear in instructorStudentListPage as well.

### CODE REVIEW

File: src/main/webapp/WEB-INF/tags/instructor/studentList.tag [hunk 1]
```diff
@@ -134,7 +134,7 @@
                       data-placement="top">
                     Edit
                   </a>
-                  <c:if test="${fromCourseDetailsPage && student.studentStatus == STUDENT_COURSE_STATUS_YET_TO_JOIN}">
+                  <c:if test="${(fromCourseDetailsPage || fromStudentListPage) && student.studentStatus == STUDENT_COURSE_STATUS_YET_TO_JOIN}">
```
Comment (Haozhe321) [inline src/main/webapp/WEB-INF/tags/instructor/studentList.tag]: This fix, whilst fixes the problem, introduces another problem. Now when instructors click on "send invite" in the Students page, the website redirects to the Courses page(and without the confirmation message that sending invite in Courses page would trigger). Can you attempt to solve the issue inside the Students page itself?

Comment (nashmia-riaz) [reply src/main/webapp/WEB-INF/tags/instructor/studentList.tag]: Yes sure! I am on it.

Comment (nashmia-riaz) [reply src/main/webapp/WEB-INF/tags/instructor/studentList.tag]: @Haozhe321 I am new to the contributors section here, so apologies for dumb questions. I have fixed the issue with redirection. However, I am having trouble showing the confirmation message. From what I've understood, the code that provides this functionality exists in instructorCourseDetails.js. Can you help me understand if this code is generated, or written?

Comment (nashmia-riaz) [reply src/main/webapp/WEB-INF/tags/instructor/studentList.tag]: @Haozhe321 Apologies for the somewhat useless commits. I am a new contributor so took some time getting used to the tests. 

I added the 2 changes you asked for. 
For fixing the redirection, I created a separated Action class for invite button in instructorStudentList. The code is very similar to the one for instructorCourseDetailsAction. 
For fixing the confirmation message, I moved the corresponding code in instructorCourseDetails.js to instructor.js ("AttachEventToSendInviteLink" - in order to avoid redundancy when using it with instructorStudentList), imported it and called it accordingly.

Can you please have a look and give me feedback? It works as far as I can tell. Thank you.

File: src/main/java/teammates/ui/controller/ActionFactory.java [hunk 2]
```diff
@@ -109,6 +109,8 @@
         map(INSTRUCTOR_STUDENT_RECORDS_PAGE, InstructorStudentRecordsPageAction.class);
         map(INSTRUCTOR_STUDENT_RECORDS_AJAX_PAGE, InstructorStudentRecordsAjaxPageAction.class);
 
+        map(INSTRUCTOR_STUDENT_LIST_REMIND, InstructorStudentListRemindAction.class);
```
Comment (Haozhe321) [inline src/main/java/teammates/ui/controller/ActionFactory.java]: This should be put above, following the alphabetical order

File: src/main/java/teammates/ui/controller/InstructorStudentListRemindAction.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            if (studentData == null) {
+                throw new EntityDoesNotExistException("Student with email " + studentEmail + " does not exist "
+                        + "in course " + courseId + "!");
+            }
+            emailDataMap.put(studentEmail,
+                    new JoinEmailData(studentData.getName(), extractStudentRegistrationKey(studentData)));
+
+            statusToUser.add(new StatusMessage(Const.StatusMessages.COURSE_REMINDER_SENT_TO + studentEmail,
+                    StatusMessageColor.SUCCESS));
+            redirectUrl = Const.ActionURIs.INSTRUCTOR_STUDENT_LIST_PAGE;
```
Comment (Haozhe321) [inline src/main/java/teammates/ui/controller/InstructorStudentListRemindAction.java]: This file is the same as InstructorCourseRemindAction.java except for line 65 and line 95. That is quite wasteful; can you think of a way use only one class and achieve both features?

File: src/main/webapp/dev/js/common/instructor.js [hunk 4]
```diff
@@ -323,6 +326,26 @@ function bindSessionDeleteLinks() {
     });
 }
 
+function attachEventToSendInviteLink() {
```
Comment (Haozhe321) [inline src/main/webapp/dev/js/common/instructor.js]: This works fine in the Course page, but doesn't work that well in the Students page. The confirmation modal is not showing up, and the entire page refreshes itself(check how it looks like in the Course remind page).

File: package.json [hunk 5]
```diff
@@ -10,7 +10,8 @@
     "lintspaces-cli": "0.6.0",
     "stylelint": "9.1.1",
     "stylelint-config-standard": "18.2.0",
-    "webpack": "3.11.0"
+    "webpack": "3.11.0",
```
[the remark this answers is not in the document]
Comment (nashmia-riaz) [reply package.json]: Honestly, I was having trouble building and testing the project via npm. It kept suggesting that I install some packages as a dependency. I apologize for the confusion and am reverting this file to the original from master (as I just realised my code doesn't need any of these dependencies to work). Thank you for pointing this out!

File: src/main/java/teammates/ui/controller/ActionFactory.java (hunk 2 above)
Comment (nashmia-riaz) [reply src/main/java/teammates/ui/controller/ActionFactory.java]: Thank you for noting this. I am fixing this as you suggested.

File: src/main/java/teammates/ui/controller/InstructorStudentListRemindAction.java (hunk 3 above)
Comment (nashmia-riaz) [reply src/main/java/teammates/ui/controller/InstructorStudentListRemindAction.java]: Yes it is. 
I can make a single class for reminding students and use it for both InstructorCourseDetails and InstructorStudentList by adding an if else check, but doing so might require changing code for both pages (deletion and addition of a series of files including studentList.tag, Data and Action files). Will that be okay?

File: src/main/webapp/dev/js/common/instructor.js (hunk 4 above)
Comment (nashmia-riaz) [reply src/main/webapp/dev/js/common/instructor.js]: It is working for me here, but I am moving towards attending to the changes you've requested above.

File: src/main/java/teammates/ui/controller/InstructorStudentListRemindAction.java (hunk 3 above)
Comment (Haozhe321) [reply src/main/java/teammates/ui/controller/InstructorStudentListRemindAction.java]: If that doesn't leave too many lines of code changes, I think that's fine :) be mindful of inheritance though because many classes inherit from Action so might have to make changes to those files as well.

Comment (nashmia-riaz) [reply src/main/java/teammates/ui/controller/InstructorStudentListRemindAction.java]: I will try to be careful of the changes I make from here on out :)

In addition to what I said above, would it be appropriate to simply use requestURL, isolate the string and check the page (if it is instructorStudentList or instructoCourseDetails)? I have noticed that it already has the link to current page as "/page/instructorCourseRemind?...." so taking the string from the last '/' to the first '?' might solve the issue as well as minimize the changes to the code files.

Comment (nashmia-riaz) [reply src/main/java/teammates/ui/controller/InstructorStudentListRemindAction.java]: Thank you so much for all the help! :)
I have a proposed solution that uses request.getHeader("referer") to get the current (i.e. page that redirects) as a string in Action. I can compare that with the ActionUri in the concerned page and redirect accordingly. Do you think that this is the right way to go? It also involves trimming and comparing the acquired current page string in the InstructorCourseRemindAction.

File: src/main/webapp/dev/js/common/instructor.js (hunk 4 above)
Comment (Haozhe321) [reply src/main/webapp/dev/js/common/instructor.js]: Yup I checked again, it's working fine!

File: src/main/java/teammates/ui/controller/InstructorStudentListRemindAction.java (hunk 3 above)
Comment (Haozhe321) [reply src/main/java/teammates/ui/controller/InstructorStudentListRemindAction.java]: The solution sounds promising, give it a try!

**Comment by Haozhe321** [comment]
@nashmia-riaz is this ready for review?

**Comment by nashmia-riaz** [comment]
@Haozhe321 Yes, sir. It is from my side.

File: src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java [hunk 6]
```diff
@@ -45,6 +45,7 @@ public void testExecuteAndPostProcess() throws Exception {
         };
 
         InstructorCourseRemindAction remindAction = getAction(submissionParams);
+        remindAction.setCurrentUrl("/page/instructorCourseEditPage?");
```
Comment (Haozhe321) [inline src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java]: This isn't necessary. This is why:
https://github.com/TEAMMATES/teammates/blob/847de291e580ffbe3f9c50b329bb29acce5818a7/src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java#L70-L85

you can verify this by deleting this added line, the test would still pass

File: src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java [hunk 7]
```diff
@@ -75,6 +76,7 @@ public void testExecuteAndPostProcess() throws Exception {
         };
 
         remindAction = getAction(submissionParams);
+        remindAction.setCurrentUrl("/page/instructorCourseDetailsPage?");
```
Comment (Haozhe321) [inline src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java]: Avoid using magic strings. Use `Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE`, same goes for the instances below

Comment (Haozhe321) [reply src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java]: Also, you have not tested that your newest addition from `INSTRUCTOR_STUDENT_LIST_PAGE` works.

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 8]
```diff
@@ -125,6 +139,22 @@ private String extractStudentRegistrationKey(StudentAttributes student) {
...
+    private String isolatePrevPageUrl() {
+        String stringTillLastSlash = currentUrl.substring(0, currentUrl.lastIndexOf('/'));
+        int indexOfSecondLastSlash = stringTillLastSlash.lastIndexOf('/');
+
+        return currentUrl.substring(indexOfSecondLastSlash, currentUrl.indexOf('?'));
+    }
+
+    private boolean compareRefererAndCurrentPage(String current) {
+        //if url is from a page and has a query, then isolate and compare
+        if (currentUrl.contains("?")) {
```
Comment (Haozhe321) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: Is the "?" necessary?

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 9]
```diff
@@ -92,7 +99,14 @@ public ActionResult execute() throws EntityDoesNotExistException {
             }
 
             statusToUser.add(new StatusMessage(Const.StatusMessages.COURSE_REMINDERS_SENT, StatusMessageColor.SUCCESS));
-            redirectUrl = Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE;
+
+            /*see and compare if the request is coming from instructorCourseDetails
+            or instructorStudentList and redirect accordingly */
+            if (compareRefererAndCurrentPage(Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE)) {
```
Comment (Haozhe321) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: I don't think the changes in this else clause is necessary. Only the `INSTRUCTOR_COURSE_DETAILS_PAGE` has a remind all students option.

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 10]
```diff
@@ -66,7 +66,14 @@ public ActionResult execute() throws EntityDoesNotExistException {
 
             statusToUser.add(new StatusMessage(Const.StatusMessages.COURSE_REMINDER_SENT_TO + studentEmail,
                                                StatusMessageColor.SUCCESS));
-            redirectUrl = Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE;
+
+            /*see and compare if the request is coming from instructorCourseDetails
+            or instructorStudentList and redirect accordingly */
+            if (compareRefererAndCurrentPage(Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE)) {
```
Comment (Haozhe321) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: the method name is confusing. It doesn't tell me it's supposed to return a boolean, even though it does return a boolean.

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 11]
```diff
@@ -125,6 +139,22 @@ private String extractStudentRegistrationKey(StudentAttributes student) {
...
     }
 
+    private String isolatePrevPageUrl() {
+        String stringTillLastSlash = currentUrl.substring(0, currentUrl.lastIndexOf('/'));
+        int indexOfSecondLastSlash = stringTillLastSlash.lastIndexOf('/');
+
+        return currentUrl.substring(indexOfSecondLastSlash, currentUrl.indexOf('?'));
+    }
+
+    private boolean compareRefererAndCurrentPage(String current) {
```
Comment (Haozhe321) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: If I'm not mistaken, the argument here shouldn't be called `current`, because it **does not** represent the current url of the page right?

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 12]
```diff
@@ -125,6 +139,22 @@ private String extractStudentRegistrationKey(StudentAttributes student) {
         return joinLink.substring(startIndex);
     }
 
+    private String isolatePrevPageUrl() {
```
Comment (Haozhe321) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: Not too sure if this is the best way. You could just pass in `Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE` or `Const.ActionURIs.INSTRUCTOR_STUDENT_LIST_PAGE` and check if the strings occur in the url.

File: src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java (hunk 6 above)
Comment (nashmia-riaz) [reply src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java]: That's true. I have removed this line now. Thank you!

File: src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java (hunk 7 above)
Comment (nashmia-riaz) [reply src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java]: My apologies. Fixing this and will keep it in mind in the future. 
Can I add test cases to the current test? If so, I will add them for InstructorCourseRemind on the INSTRUCTOR_STUDENT_LIST_PAGE right away.

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java (hunk 12 above)
Comment (nashmia-riaz) [reply src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: You are right. I didn't realize this earlier. I am implementing this like you suggested and will cater to the changes you've pointed out above in the process. Thank you. :)

File: src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java (hunk 7 above)
Comment (Haozhe321) [reply src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java]: As a general guideline, when a new feature is implemented, you should add tests to test it. So yes.

Comment (nashmia-riaz) [reply src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java]: Oh okay. I'm on it.

**Comment by nashmia-riaz** [comment]
@Haozhe321 I have added the tests and made the code simpler like you suggested. This is ready for review, but the tests weren't passing as the dependencies were down.

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 13]
```diff
@@ -92,7 +99,14 @@ public ActionResult execute() throws EntityDoesNotExistException {
             }
 
             statusToUser.add(new StatusMessage(Const.StatusMessages.COURSE_REMINDERS_SENT, StatusMessageColor.SUCCESS));
-            redirectUrl = Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE;
```
Comment (Haozhe321) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: As said, changes in this `else` clause is not necessary because only the `INSTRUCTOR_COURSE_DETAILS_PAGE` has a remind all option.

File: src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java [hunk 14]
```diff
@@ -120,6 +138,7 @@ public void testExecuteAndPostProcess() throws Exception {
                 Const.ParamsNames.COURSE_ID, courseId
         };
         remindAction = getAction(addUserIdToParams(instructorId, submissionParams));
+        remindAction.setCurrentUrl(Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE);
```
Comment (Haozhe321) [inline src/test/java/teammates/test/cases/action/InstructorCourseRemindActionTest.java]: This line and line 178 are not necessary. These test cases refer to the remind all option from INSTRUCTOR_COURSE_DETAILS_PAGE I believe, so there's no need to explicitly set the url.

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 15]
```diff
@@ -125,6 +139,11 @@ private String extractStudentRegistrationKey(StudentAttributes student) {
         return joinLink.substring(startIndex);
     }
 
+    private boolean isRequestedFromPage(String requestorUrl) {
+        //if Url contains the requestor's url, then return true
```
Comment (Haozhe321) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: The method is pretty self-explanatory, so this comment can be removed

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 16]
```diff
@@ -125,6 +139,11 @@ private String extractStudentRegistrationKey(StudentAttributes student) {
         return joinLink.substring(startIndex);
     }
 
+    private boolean isRequestedFromPage(String requestorUrl) {
```
Comment (Haozhe321) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: I wouldn't call the argument `requestorUrl`, because this method is verifying if the request url contains a certain `Const.ActionURIs`. Perhaps something alongs the lines of `UrlToVerify` is more suitable. `requestorUrl` implies that it is the requesting url, which isn't what you are trying to say.

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 17]
```diff
@@ -66,7 +66,14 @@ public ActionResult execute() throws EntityDoesNotExistException {
 
             statusToUser.add(new StatusMessage(Const.StatusMessages.COURSE_REMINDER_SENT_TO + studentEmail,
                                                StatusMessageColor.SUCCESS));
-            redirectUrl = Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE;
+
+            /*see and compare if the request is coming from instructorCourseDetails
```
Comment (Haozhe321) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: This comment can be removed as well.

**Comment by nashmia-riaz** [comment]
@Haozhe321 Thank you for taking out time to point out the changes once again :)
This is ready for review.

**Comment by Haozhe321** [comment]
LGTM! @wkurniawan07 pls have a look

File: src/main/java/teammates/ui/controller/Action.java [hunk 18]
```diff
@@ -102,6 +105,7 @@ public void init(HttpServletRequest req) {
     protected void initialiseAttributes(HttpServletRequest req) {
         request = req;
         requestUrl = HttpRequestHelper.getRequestedUrl(request);
+        currentUrl = request.getHeader("referer");
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/Action.java]: Why `currentUrl` when `referer` (or rather, `referrer`) can make a better variable name?

File: src/main/java/teammates/ui/controller/ActionFactory.java [hunk 19]
```diff
@@ -159,7 +159,7 @@ private static Action getAction(String uri) {
             return controllerClass.newInstance();
         } catch (Exception e) {
             Assumption.fail("Could not create the action for " + uri + ": "
-                            + TeammatesException.toStringWithStackTrace(e));
+                    + TeammatesException.toStringWithStackTrace(e));
```
Comment (wkurniawan07) [inline src/main/java/teammates/ui/controller/ActionFactory.java]: Undo this change?

Comment (nashmia-riaz) [reply src/main/java/teammates/ui/controller/ActionFactory.java]: Sure!

File: src/main/java/teammates/ui/controller/Action.java (hunk 18 above)
Comment (nashmia-riaz) [reply src/main/java/teammates/ui/controller/Action.java]: That does sound better! Changing this to referrer.

Comment (whipermr5) [reply src/main/java/teammates/ui/controller/Action.java]: Can use the new variable in `validateOriginIfRequired`.

**Comment by nashmia-riaz** [comment]
@wkurniawan07 I have pushed the minor changes you requested, but the build is suddenly failing and I'm not sure why. The errors occur in files I have not modified in anyway. Can you please take a look and let me know what I can do to fix this?

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 20]
```diff
@@ -66,7 +66,12 @@ public ActionResult execute() throws EntityDoesNotExistException {
 
             statusToUser.add(new StatusMessage(Const.StatusMessages.COURSE_REMINDER_SENT_TO + studentEmail,
                                                StatusMessageColor.SUCCESS));
-            redirectUrl = Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE;
+
+            if (isRequestedFromPage(Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE)) {
+                redirectUrl = Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE;
+            } else if (isRequestedFromPage(Const.ActionURIs.INSTRUCTOR_STUDENT_LIST_PAGE)) {
+                redirectUrl = Const.ActionURIs.INSTRUCTOR_STUDENT_LIST_PAGE;
+            }
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: What happens if the HTTP referrer is disabled on the user's browser?

Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: It would be better to use a request parameter to indicate where to redirect to instead of relying on the referrer.

Comment (nashmia-riaz) [reply src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: @whipermr5 Thank you, sir. 👍 
I didn't consider this before but it does sound much more stable. I am working on this now.

**Comment by nashmia-riaz** [comment]
@whipermr5 Sorry for the delay. This is ready for review. :)

File: src/main/java/teammates/ui/template/StudentListStudentData.java [hunk 21]
```diff
@@ -30,25 +30,26 @@ public StudentListStudentData(String googleId, String studentName, String studen
...
-                                                                           course, studentEmail, googleId);
+                                                                           course, studentEmail, googleId, null);
         this.courseStudentRecordsLink = furnishLinkWithCourseEmailAndUserId(Const.ActionURIs.INSTRUCTOR_STUDENT_RECORDS_PAGE,
-                                                                            course, studentEmail, googleId);
+                                                                            course, studentEmail, googleId, null);
     }
 
     private String furnishLinkWithCourseEmailAndUserId(String rawLink, String course, String studentEmail,
-                                                       String googleId) {
+                                                       String googleId, String previousPage) {
```
Comment (whipermr5) [inline src/main/java/teammates/ui/template/StudentListStudentData.java]: Instead of modifying this method, simply modify `rawLink` before passing it here.

**Comment by nashmia-riaz** [comment]
@whipermr5 this is ready for review from my side.

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 22]
```diff
@@ -66,7 +68,12 @@ public ActionResult execute() throws EntityDoesNotExistException {
 
             statusToUser.add(new StatusMessage(Const.StatusMessages.COURSE_REMINDER_SENT_TO + studentEmail,
                                                StatusMessageColor.SUCCESS));
-            redirectUrl = Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE;
+
+            if (isRequestedFromPage(previousPage, Const.PageNames.INSTRUCTOR_COURSE_DETAILS_PAGE)) {
+                redirectUrl = Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE;
+            } else if (isRequestedFromPage(previousPage, Const.PageNames.INSTRUCTOR_STUDENT_LIST_PAGE)) {
+                redirectUrl = Const.ActionURIs.INSTRUCTOR_STUDENT_LIST_PAGE;
+            }
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: It'll be better to default to one of the cases, in case both conditions fail.

**Comment by nashmia-riaz** [comment]
@whipermr5 How's this? I assumed any case as default case would be good, so went with the student list page.

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 23]
```diff
@@ -125,6 +132,10 @@ private String extractStudentRegistrationKey(StudentAttributes student) {
         return joinLink.substring(startIndex);
     }
 
+    private boolean isRequestedFromPage(String urlToVerify, String urlToCompareWith) {
+        return urlToVerify.equals(urlToCompareWith);
+    }
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: I think this method can be removed; it doesn't add much clarity.

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java [hunk 24]
```diff
@@ -66,7 +68,12 @@ public ActionResult execute() throws EntityDoesNotExistException {
 
             statusToUser.add(new StatusMessage(Const.StatusMessages.COURSE_REMINDER_SENT_TO + studentEmail,
                                                StatusMessageColor.SUCCESS));
-            redirectUrl = Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE;
+
+            if (isRequestedFromPage(previousPage, Const.PageNames.INSTRUCTOR_COURSE_DETAILS_PAGE)) {
```
Comment (whipermr5) [inline src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: Use `Const.PageNames.INSTRUCTOR_COURSE_DETAILS_PAGE.equals(previousPage)` instead as `previousPage` might be null.

File: src/main/java/teammates/ui/template/StudentListTeamData.java [hunk 25]
```diff
@@ -13,14 +13,12 @@
     private List<StudentListStudentData> students;
 
     public StudentListTeamData(TeamDetailsBundle team, Map<String, String> emailPhotoUrlMapping, String googleId,
-                               String sessionToken) {
+                               String sessionToken, String previousPage) {
```
Comment (whipermr5) [inline src/main/java/teammates/ui/template/StudentListTeamData.java]: Use 8-space indentation

File: src/main/webapp/dev/js/main/instructorStudentList.js [hunk 26]
```diff
@@ -536,6 +538,5 @@ $(document).ready(() => {
             toggleSort($(this));
         }
     });
-
```
Comment (whipermr5) [inline src/main/webapp/dev/js/main/instructorStudentList.js]: This is an unrelated change

File: src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java (hunk 24 above)
Comment (whipermr5) [reply src/main/java/teammates/ui/controller/InstructorCourseRemindAction.java]: To make it clear, can extract a local variable:
```java
boolean isRequestedFromCourseDetailsPage = Const.PageNames.INSTRUCTOR_COURSE_DETAILS_PAGE.equals(previousPage);
redirectUrl = isRequestedFromCourseDetailsPage ? Const.ActionURIs.INSTRUCTOR_COURSE_DETAILS_PAGE
                                               : Const.ActionURIs.INSTRUCTOR_STUDENT_LIST_PAGE;
```

File: src/main/java/teammates/ui/template/StudentListTeamData.java (hunk 25 above)
Comment (nashmia-riaz) [reply src/main/java/teammates/ui/template/StudentListTeamData.java]: I have to point out that I have noticed inconsistency with the 8-space indentation rule in already written code. I must ask out of genuine curiosity, is there a reason for this that I'm missing?
